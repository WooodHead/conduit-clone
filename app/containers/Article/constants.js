/*
 *
 * Article constants
 *
 */

export const DEFAULT_ACTION = 'app/Article/DEFAULT_ACTION';
export const FETCH_ARTICLE = 'app/Article/FETCH_ARTICLE';
export const ARTICLE_LOADED = 'app/Article/ARTICLE_LOADED';
export const LOAD_ARTICLE_ERROR = 'app/Article/LOAD_ARTICLE_ERROR';
