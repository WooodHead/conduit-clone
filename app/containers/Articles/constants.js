/*
 *
 * Articles constants
 *
 */

export const FETCH_ARTICLES = 'app/Articles/FETCH_ARTICLES';
export const ARTICLES_LOADED = 'app/Articles/ARTICLES_LOADED';
export const LOAD_ARTICLES_ERROR = 'app/Articles/LOAD_ARTICLES_ERROR';
