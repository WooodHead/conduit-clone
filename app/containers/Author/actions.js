/*
 *
 * Author actions
 *
 */

import {
  DEFAULT_ACTION,
  FETCH_AUTHOR_PROFILE,
  AUTHOR_PROFILE_LOADED,
  AUTHOR_PROFILE_LOADING_ERROR,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function fetchAuthorProfile(username) {
  return {
    type: FETCH_AUTHOR_PROFILE,
    username,
  };
}

export function authorProfileLoaded(profile) {
  return {
    type: AUTHOR_PROFILE_LOADED,
    profile,
  };
}

export function authorProfileLoadingError() {
  return {
    type: AUTHOR_PROFILE_LOADING_ERROR,
  };
}
