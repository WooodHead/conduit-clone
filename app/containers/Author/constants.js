/*
 *
 * Author constants
 *
 */

export const DEFAULT_ACTION = 'app/Author/DEFAULT_ACTION';
export const FETCH_AUTHOR_PROFILE = 'app/Author/FETCH_AUTHOR_PROFILE';
export const AUTHOR_PROFILE_LOADED = 'app/Author/AUTHOR_PROFILE_LOADED';
export const AUTHOR_PROFILE_LOADING_ERROR = 'app/Author/AUTHOR_PROFILE_LOADING_ERROR';
