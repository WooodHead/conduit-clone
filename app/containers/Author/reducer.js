/*
 *
 * Author reducer
 *
 */

import { fromJS } from 'immutable';
import {
  AUTHOR_PROFILE_LOADED,
  AUTHOR_PROFILE_LOADING_ERROR,
} from './constants';

const initialState = fromJS({
  author: {
    fetching: true,
    error: false,
    data: {},
  },
});

function authorReducer(state = initialState, action) {
  switch (action.type) {
    case AUTHOR_PROFILE_LOADED:
      return state
        .setIn(['author', 'fetching'], false)
        .setIn(['author', 'data'], fromJS(action.profile));
    case AUTHOR_PROFILE_LOADING_ERROR:
      return state
        .setIn(['author', 'fetching'], false)
        .setIn(['author', 'error'], true);
    default:
      return state;
  }
}

export default authorReducer;
