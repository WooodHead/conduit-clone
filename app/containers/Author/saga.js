import { takeLatest, call, put } from 'redux-saga/effects';
import {
  FETCH_AUTHOR_PROFILE,
} from 'containers/Author/constants';
import {
  authorProfileLoaded,
  authorProfileLoadingError,
} from 'containers/Author/actions';
import request from 'utils/request';
import { removeFirstCharacter } from 'utils/string';

const API_DOMAIN = 'https://conduit.productionready.io/api';

export function* getAuthorProfile(param) {
  const { username } = param;
  const PROFILE_ENDPOINT = `${API_DOMAIN}/profiles/${removeFirstCharacter('@', username)}`;
  try {
    const payload = yield call(request, PROFILE_ENDPOINT);
    yield put(authorProfileLoaded(payload.profile));
  } catch (err) {
    yield put(authorProfileLoadingError());
  }
}

export default function* authorProfileData() {
  yield takeLatest(FETCH_AUTHOR_PROFILE, getAuthorProfile);
}
