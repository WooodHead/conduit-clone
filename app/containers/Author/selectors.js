import { createSelector } from 'reselect';

/**
 * Direct selector to the author state domain
 */
const selectAuthorDomain = (state) => state.get('author');

/**
 * Default selector used by Author
 */
const makeSelectAuthorData = () => createSelector(
  selectAuthorDomain,
  (substate) => substate.getIn(['author', 'data']).toJS()
);

const makeSelectAuthorFetching = () => createSelector(
  selectAuthorDomain,
  (substate) => substate.getIn(['author', 'fetching'])
);

const makeSelectAuthorError = () => createSelector(
  selectAuthorDomain,
  (substate) => substate.getIn(['author', 'error'])
);

export {
  selectAuthorDomain,
  makeSelectAuthorData,
  makeSelectAuthorFetching,
  makeSelectAuthorError,
};
