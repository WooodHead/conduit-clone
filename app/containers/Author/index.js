/**
 *
 * Author
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { isEmpty } from 'lodash';

import {
  Container,
  Alert,
} from 'reactstrap';

import Loader from 'components/Loader';
import AuthorBanner from 'components/AuthorBanner';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import {
  makeSelectAuthorData,
  makeSelectAuthorFetching,
  makeSelectAuthorError,
} from './selectors';
import reducer from './reducer';
import saga from './saga';
import { fetchAuthorProfile } from './actions';
import { AuthorPropTypes } from '../../PropTypesValues';

export class Author extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  componentDidMount() {
    this.props.onFetchAuthorProfile(this.props.match.params.username);
  }

  render() {
    const {
      fetching,
      error,
      data,
    } = this.props.author;
    let content;
    if (!fetching) {
      if (!error && !isEmpty(data)) {
        const { bio, following, image, username } = data;
        content = (
          <div>
            <AuthorBanner
              bio={bio}
              following={following}
              image={image}
              username={username}
            />
          </div>
        );
      } else {
        content = (
          <Container>
            <Alert color="danger">
              Something went wrong.
            </Alert>
          </Container>
        );
      }
    } else {
      content = (
        <Container>
          <Loader />
        </Container>
      );
    }

    return (
      <div>
        <Helmet>
          <title>Author</title>
          <meta name="description" content="Description of Author" />
        </Helmet>
        {content}
      </div>
    );
  }
}

Author.propTypes = {
  onFetchAuthorProfile: PropTypes.func.isRequired,
  author: PropTypes.shape({
    fetching: PropTypes.bool,
    error: PropTypes.bool,
    data: AuthorPropTypes,
  }),
  match: PropTypes.shape({
    params: PropTypes.shape({
      username: PropTypes.string.isRequired,
    }),
  }),
};

const mapStateToProps = createStructuredSelector({
  author: createStructuredSelector({
    fetching: makeSelectAuthorFetching(),
    error: makeSelectAuthorError(),
    data: makeSelectAuthorData(),
  }),
});

export function mapDispatchToProps(dispatch) {
  return {
    onFetchAuthorProfile: (username) => {
      dispatch(fetchAuthorProfile(username));
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'author', reducer });
const withSaga = injectSaga({ key: 'author', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(Author);
