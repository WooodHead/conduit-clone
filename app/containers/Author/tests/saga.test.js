/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import { FETCH_AUTHOR_PROFILE } from 'containers/Author/constants';

import {
  authorProfileLoadingError,
} from 'containers/Author/actions';

import authorProfileData, { getAuthorProfile } from '../saga';

const response = {};

describe('defaultSaga Saga', () => {
  let getAuthorProfileGenerator;

  beforeEach(() => {
    getAuthorProfileGenerator = getAuthorProfile({
      username: '@john_doe',
    });

    const selectDescriptor = getAuthorProfileGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getAuthorProfileGenerator.next(response).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the authorProfileLoadingError action if it requests throws an error', () => {
    const putDescriptor = getAuthorProfileGenerator.throw().value;
    expect(putDescriptor).toEqual(put(authorProfileLoadingError()));
  });
});

describe('authorProfileData Saga', () => {
  const authorProfile = authorProfileData();

  it('should start task to watch for FETCH_AUTHOR_PROFILE action', () => {
    const takeLatestDescriptor = authorProfile.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(FETCH_AUTHOR_PROFILE, getAuthorProfile));
  });
});
