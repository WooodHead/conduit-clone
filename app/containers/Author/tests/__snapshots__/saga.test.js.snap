// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultSaga Saga should dispatch the authorProfileLoadingError action if it requests throws an error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://conduit.productionready.io/api/profiles/john_doe",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`defaultSaga Saga should dispatch the authorProfileLoadingError action if it requests throws an error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "profile": undefined,
      "type": "app/Author/AUTHOR_PROFILE_LOADED",
    },
    "channel": null,
  },
}
`;
